### ISSUE TITLE
Contextual token values

### ISSUE THREAD

**Issue by ilikescience** [design-tokens/community-group#204]
> **Note**
> This is meant to be a catchall issue for a number of separate conversations around contextual values. See #188, #169, #135, and #2 for previous conversations

Reading back through issues, it seems like we should make three interlinked decisions soon:

1. Should the spec support contextual token values? (aka theming, fluid values, conditional values)
    a. If so, what contexts should be supported?
    b. And what should the context be?

My current thinking:

## Should the spec support contextual values?
Yes. Changing a design based on context is a major use case for tokens, so it would be good to include it in the spec. 

## If so, what contexts should be supported?
Those that are independent of platforms. For example (not exhaustive):
- Platform-independent:
    - Viewport width
    - Accessibility preferences (high contrast, reduced motion)
    - Color theme
- Platform-dependent:
    - version or environment (ie CSS @supports or iOS version)
    - unique units like CSS's lvh, dvb

## And what should the syntax be?
@romainmenke 's initial proposal is a good place to start, though I'd recommend calling them "contextual" values instead of "conditional" values.

```json
{
  "backgroundColor": {
    "$type": "color",
    "$value": "cyan",
    "$contextualValues": [
      {
        "$context": "width",
        "$condition": ">500px",
        "$value": "purple"
      },
      {
        "$context": "colorScheme",
        "$condition": "dark",
        "$value": "yellow"
      }
    ]
  }
}
```

Though an argument could be made for grouping token mappings under contexts like so:

```json
{
  "backgroundColor": {
    "$type": "color",
    "default": {
      "$value": "cyan"
    },
    "context": {
      "wide": {
        "$value": "purple"
      },
      "dark": {
        "$value": "yellow"
      }
    }
  },
  "$contexts": {
    "width": [
      {
        "$condition": ">500px",
        "$contextualValues": {
          "backgroundColor-default": "{backgroundColor-context-wide}"
        }
      }
    ],
    "colorScheme": [
      {
        "$condition": "dark",
        "contextualValues": {
          "backgroundColor-default": "{backgroundColor-context-dark}"
        }
      }
    ]
  }
}
```

I'm using aliases in the second example to demonstrate how contexts could be additive to grouping.

---

Overall, I think that we can define a set of contexts that are supported by all target platforms, then define a specification for how the syntax in the token file should be interpreted in each language.

**Comment by romainmenke** [comment]
> - version or environment (ie CSS https://github.com/supports or iOS version)
> - unique units like CSS's lvh, dvb

I don't think there is much value **at this** time in supporting platform specific contexts and that it will be problematic to implement in design tools.

There is just no way design tools can provide meaningful and useful UI for all the platform specific contexts of all the possible platforms.

Keeping the scope small is key here, I think.

**Comment by ilikescience** [comment]
> I don't think there is much value at this time in supporting platform specific contexts and that it will be problematic to implement in design tools.

Agreed!

**Comment by kevinmpowell** [comment]
Related to #193

**Comment by c1rrus** [comment]
I agree it would be great to have a mechanism in the format for this kind of thing, but I wonder whether the spec needs to pre-define contexts at all.

What if we let people provide any number of alternative values for a given token, each with a user-defined identifier. For example:

```json
{
  "mighty-morphing-token": {
    "$type": "dimension",
    "$value": "16px",
    "$altValues": {
      "high-info-density": "12px",
      "low-info-density": "20px",
      "random-thing": "17px"
    }
  }
}
```

From a tooling perspective, the only requirement would be to let the user to provide one (or more?) of their chosen keys and then resolve token values to the corresponding `$altValue` if it exists. Where non exists, `$value` is used as a fallback.

So the above token's value would be `16px` by default, but if the user has asked the tool to prefer `"low-info-density"` values, ot would resolve to `20px` instead.

### Example 1: Translation tool

A translation tool like Style Dictionary could let the user pass in desired alt value keys via a command-line option or config file.

Let's imagine we have such a tool, usage could look something like this:

```sh
$ dtcg2css input.tokens.json > output.css
```

Might use default values and produce something like:

```css
:root {
  --mighty-morphing-token: 16px;
}
```

But running it like this:

```sh
$ dtcg2css --alt-value=random-thing input.tokens.json > output-random.css
```

...would produce:

```css
:root {
  --mighty-morphing-token: 17px;
}
```

### Example 2: UI design tool

A design tool like Figma might present a list of all alt value keys in a `.tokens` file to the user and let them pick which one(s) to apply.

Wherever they've used those tokens in their designs, the design would update to show the currently applied settings.

I'm imagining the experience to be similar to how you can [switch "token sets" on and off in the Tokens Studio Figma plug-in](https://docs.tokens.studio/themes/token-sets).

**Comment by romainmenke** [comment]
> I agree it would be great to have a mechanism in the format for this kind of thing, but I wonder whether the spec needs to pre-define contexts at all.
>
>What if we let people provide any number of alternative values for a given token, each with a user-defined identifier. For example:

I fear this would not be useful.
I also don't see how it different from regular groups.

Because this is all user defined we can not add any logic for it in translation tools.
It would be extra complexity without any extra features.

**Comment by VIPofJoy** [comment]
I have experienced the magic of this plugin, it is very friendly to support multiple brands and multiple context functions within a brand.
https://user-images.githubusercontent.com/121750728/229291388-a4a21370-55e2-44c2-b280-5f29235dae62.mov
